### PR TITLE
[EZEE_2772] Calendar customization

### PR DESCRIPTION
**Pull request by michalmilc90**
| Question      | Answer
| ------------- | ---
| JIRA Ticket   | [EZEE-2772](https://jira.ez.no/browse/EZEE-2772)
| Versions      | master (3.0)

Related PRs:

- [x] https://github.com/ezsystems/ezplatform-calendar/pull/6
- [x] https://github.com/ezsystems/ezplatform-calendar/pull/10
- [x] https://github.com/ezsystems/ezplatform-calendar/pull/19
- [x] https://github.com/ezsystems/ezplatform-calendar/pull/26
- [ ] https://github.com/ezsystems/ezplatform-calendar/pull/40

### CODE REVIEW

File: docs/guide/customizing_calendar.md [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+Optionally, you can [change the colors and icons](#customizing-colors-and-icons) and make the widget look differently depending on the [SiteAccess configuration](../guide/siteaccess.md#configuring-siteaccesses).
+
+## Configuring custom calendar events
+
+Configuring custom events allows you to tailor the calendar to your needs.
+You can create a custom event by taking a few steps to define the event and actions for it.
+
+First, create a new event by creating `src\Calendar\EventType\MyEvent.php`
+
+``` PHP hl_lines="7"
```
Comment (DominikaK) [inline docs/guide/customizing_calendar.md]:
```suggestion
``` php hl_lines="7"
```
(doesn't matter for rendering, but for consistency)

File: docs/guide/customizing_calendar.md [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+final class MyEvent extends Event
+{
+    public function __construct(string $id, DateTimeInterface $dateTime, MyEventType $type)
+    {
+        parent::__construct($type, $id, $dateTime);
+    }
+}
+```
+
+Here, you define a new class for you event based on `\EzSystems\EzPlatformCalendar\Calendar\Event`.
```
Comment (DominikaK) [inline docs/guide/customizing_calendar.md]:
```suggestion
Here, you define a new class for your event based on `\EzSystems\EzPlatformCalendar\Calendar\Event`.
```

File: docs/guide/customizing_calendar.md [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+Line 7 points to the custom event definition for actions associated with your event.
+
+Proceed with providing the definition for your event by creating `\App\Calendar\EventType\MyEventType.php`:
+
+``` PHP hl_lines="28 29 30 31 33 34 35 36"
```
Comment (DominikaK) [inline docs/guide/customizing_calendar.md]:
```suggestion
``` php hl_lines="28 29 30 31 33 34 35 36"
```

File: docs/guide/customizing_calendar.md [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+    }
+    
+    public function getTypeLabel(): string
+        {
+            return $this->translator->trans('event_type.private.label', [], 'my_calendar_extension');
+        }
+}
+```
+Here, lines 28-31 are responsible for building names for your custom events using a pattern.
+Lines 33-36 use the `\EzSystems\EzPlatformCalendar\Calendar\EventType\EventTypeInterface::getTypeLabel` to generate a label.
```
Comment (DominikaK) [inline docs/guide/customizing_calendar.md]:
```suggestion
Lines 33-36 use `\EzSystems\EzPlatformCalendar\Calendar\EventType\EventTypeInterface::getTypeLabel` to generate a label.
```

File: docs/guide/customizing_calendar.md [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+
+Complete the procedure by registering the new event:
+
+``` YAML
```
Comment (DominikaK) [inline docs/guide/customizing_calendar.md]:
```suggestion
``` yaml
```

File: docs/guide/customizing_calendar.md [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+services:
+
```
Comment (DominikaK) [inline docs/guide/customizing_calendar.md]:
```suggestion
```

File: docs/guide/customizing_calendar.md [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+  App\Calendar\EventType\CustomEventType:
+        tags:
+            - { name: ezplatform.calendar.event_type }
+
+```
+
+### Adding actions to events
+
+The definition of the event action implements [`Calendar\EventAction\EventActionInterface`](https://github.com/ezsystems/ezplatform-calendar/blob/master/src/lib/Calendar/EventAction/EventActionInterface.php).
+The [`EventTypeInterface::getActions`](https://github.com/ezsystems/ezplatform-calendar/blob/b88392263deec46d7603d7ecf67ab9bb908787e0/src/lib/Calendar/EventType/EventTypeInterface.php#L44)
```
Comment (DominikaK) [inline docs/guide/customizing_calendar.md]:
```suggestion
The [`EventTypeInterface::getActions`](https://github.com/ezsystems/ezplatform-calendar/blob/master/src/lib/Calendar/EventType/EventTypeInterface.php#L44) 
```

File: docs/guide/customizing_calendar.md [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+        tags:
+            - { name: ezplatform.calendar.event_type }
+
+```
+
+### Adding actions to events
+
+The definition of the event action implements [`Calendar\EventAction\EventActionInterface`](https://github.com/ezsystems/ezplatform-calendar/blob/master/src/lib/Calendar/EventAction/EventActionInterface.php).
+The [`EventTypeInterface::getActions`](https://github.com/ezsystems/ezplatform-calendar/blob/b88392263deec46d7603d7ecf67ab9bb908787e0/src/lib/Calendar/EventType/EventTypeInterface.php#L44) 
+method specifies actions supported by a specific even type.
```
Comment (DominikaK) [inline docs/guide/customizing_calendar.md]:
```suggestion
method specifies actions supported by a specific event type.
```

File: docs/guide/customizing_calendar.md [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+
+You can add an action to an event by injecting it using the event type constructor:
+
+``` PHP
```
Comment (DominikaK) [inline docs/guide/customizing_calendar.md]:
```suggestion
``` php
```

File: docs/guide/customizing_calendar.md [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+<?php
+
+# ...
```
Comment (DominikaK) [inline docs/guide/customizing_calendar.md]:
```suggestion
// ...
```
AFAIK that's the standard for comments in PHP

File: docs/guide/customizing_calendar.md [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+{
+    /** @var \EzSystems\EzPlatformCalendar\Calendar\EventAction\EventActionCollection */
+    private $actions;
+
+    public function __construct(iterable $actions)
+    {
+        $this->actions = new EventActionCollection($actions);
+    }
+
+    # ...
```
Comment (DominikaK) [inline docs/guide/customizing_calendar.md]:
```suggestion
    // ...
```

File: docs/guide/customizing_calendar.md [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+    public function getActions(): EventActionCollection
+    {
+        return $this->actions;
+    }
+}
+```
+
+The added actions require registering them as services:
+
+``` YAML hl_lines="4 7 10 15"
```
Comment (DominikaK) [inline docs/guide/customizing_calendar.md]:
```suggestion
``` yaml hl_lines="4 7 10 15"
```

File: docs/guide/customizing_calendar.md [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+## Configuring custom calendar event sources
+
+You can implement a custom event source by using:
+
+- [`EzSystems\EzPlatformCalendar\Calendar\EventSourceInterface`](https://github.com/ezsystems/ezplatform-calendar/blob/master/src/lib/Calendar/EventSource/EventSourceInterface.php)
+- [`EzSystems\EzPlatformCalendar\Calendar\InMemoryEventSource`](https://github.com/ezsystems/ezplatform-calendar/blob/master/src/lib/Calendar/EventSource/InMemoryEventSource.php)
+
+To add an in-memory collection as an event source, create `app/Calendar/EventSourceFactory/MyEventSourceFactory.php`:
+
+``` PHP hl_lines="26 27 28"
```
Comment (DominikaK) [inline docs/guide/customizing_calendar.md]:
```suggestion
``` php hl_lines="26 27 28"
```

File: docs/guide/customizing_calendar.md [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+}
+```
+
+!!! note
+
+    When providing the list of events for  `$collection_name = new EventCollection()`, you must put all the `createEvent()` elements sorted according to their time stamp.
+    
+    For example:
+    
+    ```PHP
```
Comment (DominikaK) [inline docs/guide/customizing_calendar.md]:
```suggestion
    ```php
```

File: docs/guide/customizing_calendar.md [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+    $collection_name = new EventCollection([
+        $this->createEvent("Event 1", new DateTime("2019-01-01")),
+        $this->createEvent("Event 2", new DateTime("2019-01-02")),
+        // ...
+    ```    
+
+Complete the procedure by registering the new source:
+
+``` YAML
```
Comment (DominikaK) [inline docs/guide/customizing_calendar.md]:
```suggestion
``` yaml
```

File: docs/guide/customizing_calendar.md [hunk 16]
```diff
@@ -0,0 +1,257 @@
...
+```
+
+## Customizing colors and icons
+
+You can change the foreground and background color of a custom event or change the icon of an event or action.
+The setting is SiteAccess-aware.
+
+To customize the appearance settings, add the following configuration to `config/packages/ezplatform.yml`:
+
+``` YAML hl_lines="6"
```
Comment (DominikaK) [inline docs/guide/customizing_calendar.md]:
```suggestion
``` yaml hl_lines="6"
```

File: mkdocs.yml [hunk 17]
```diff
@@ -139,6 +139,7 @@ pages:
             - 'Creating custom Page blocks': 'guide/extending_page.md'
             - 'Creating custom richtext blocks': 'guide/richtext_block.md'
             - 'Extending Form Builder': 'guide/extending_form_builder.md'
+            - 'Customizing Calendar widget': 'guide/customizing_calendar.md'
```
Comment (DominikaK) [inline mkdocs.yml]:
```suggestion
            - 'Customizing calendar widget': 'guide/customizing_calendar.md'
```

File: docs/guide/extending/customizing_calendar.md [hunk 18]
```diff
@@ -0,0 +1,261 @@
...
+``` yaml
+services:
+
+    App\Calendar\EventSourceFactory\MyEventSourceFactory:
+        arguments:
+            $eventType: '@App\Calendar\EventType\MyEventType'
+    
+    App\Calendar\EventSource\HolidaysEventSource:
+        class: EzSystems\EzPlatformCalendar\Calendar\EventSource\InMemoryEventSource
+        factory: '@App\Calendar\EventSourceFactory\MyEventSourceFactory','createEventSource']
```
Comment (bdunogier) [inline docs/guide/extending/customizing_calendar.md]: Missing opening square bracket:

```suggestion
        factory: ['@App\Calendar\EventSourceFactory\MyEventSourceFactory','createEventSource']
```

**Comment by MagdalenaZuba** [comment]
@bdunogier  please note that this PR is still labeled as "Work in progress"

**Comment by bdunogier** [comment]
No problem @MagdalenaZuba. I admit that I didn't look into the details. I'm customizing the calendar for today's sales enablement demo, and chose to report what I ran into. With what I mentioned + the current WIP version, I got my events to show up !

**Comment by DominikaK** [comment]
Closing in favour of https://github.com/ezsystems/developer-documentation/pull/993 with a cleaner history

**Pull request closed without merging**